### PR TITLE
refactor: add `index` method for `StacksEpochId`

### PR DESCRIPTION
**Pull request by obycode**
This allows us to cleanup some magic numbers when accessing epochs in a list and make the expected behavior more clear to the reader.

### CODE REVIEW

**Review by jcnelson — DISMISSED** [review]
LGTM assuming CI passes

**Comment by obycode** [comment]
There was once place acting on a list of epochs where the list was incomplete, so the standard indices shouldn't be used. This fix should resolve the failing tests.

**Comment by kantai** [comment]
I think it'd be good to try to clean up the magic numbers and accessors used in epoch lists, but I'm kind of wary of encoding the indexes, even though it probably is perfectly safe (because there's no legal epoch ordering where `2.4` doesn't exist, but `2.5` does).

An alternative that I think would have some of the same benefits would be a wrapper struct...

```rust
pub struct EpochList (Vec<StacksEpoch>)
```

And then defining:

```rust
impl Index<&StacksEpochId> for EpochList {
  type Output = StacksEpoch;
  fn index(&self, index: &StacksEpochId) -> &StacksEpoch {
     &self[StacksEpoch::find_by_epoch_id(self, index).unwrap()]
  }
}
```

**Comment by obycode** [comment]
Sure, that does seem nice. Likely a little slower, but I don't think we'd use it anywhere that is performance critical.

**Comment by obycode** [comment]
Ok, re-refactored with `pub struct EpochList (Vec<StacksEpoch>)`

File: stackslib/src/core/mod.rs
Comment (kantai) [inline stackslib/src/core/mod.rs:1]: Does it make sense to have the `STACKS_EPOCHS_MAINNET`, `_TESTNET` and `_REGTEST` lazy statics be `EpochList`s instead? Or would that generate way too many changes?

File: stacks-common/src/types/mod.rs [hunk 1]
```diff
@@ -460,3 +461,89 @@ impl<L: PartialEq + Eq> Ord for StacksEpoch<L> {
...
+
+    /// Determine which epoch, if any, a given burnchain height falls into.
+    pub fn epoch_at_height(&self, height: u64) -> Option<StacksEpoch<L>> {
+        for epoch in self.0.iter() {
+            if epoch.start_height <= height && height < epoch.end_height {
+                return Some(epoch.clone());
+            }
+        }
+        None
+    }
```
Comment (kantai) [inline stacks-common/src/types/mod.rs:508]: I think these should invoke `StacksEpoch::find_epoch()` so that we avoid repeating the height searching logic.

**Review by kantai — COMMENTED** [review]
This looks good to me-- it definitely improves the legibility of the epochs lists! Just a couple comments.

File: stackslib/src/core/mod.rs
Comment (obycode) [reply stackslib/src/core/mod.rs:1]: Changed in e9c141f5220feea9a8f5a1b5bdce554fb4d44e74.

File: stacks-common/src/types/mod.rs (hunk 1 above)
Comment (obycode) [reply stacks-common/src/types/mod.rs]: Done in e9c141f5220feea9a8f5a1b5bdce554fb4d44e74.

**Review by kantai — APPROVED** [review]

**Review by kantai — APPROVED** [review]

**Review by jcnelson — APPROVED** [review]

**Pull request merged**